dwh_design_conventions.md

```diff
@@ -0,0 +1,177 @@
+thinking about data warehouse design, working in concert with our dbt-driven 
+modeling [best practices](https://docs.getdbt.com/docs/best-practices). 
+The [coding conventions](https://github.com/fishtown-analytics/corp/blob/master/dbt_coding_conventions.md)
+are still as true as ever; we're applying the same thinking 
+beyond the `select` statement, to the repo and warehouse levels.
+
+### Note on Custom Schemata
+
+The concept of a folder hierarchy is standard; the organization of views and
+tables within a database schema, less so. dbt offers a lot by way of 
+[custom schema names](https://docs.getdbt.com/docs/using-custom-schemas), and
+soon [model aliasing](https://github.com/fishtown-analytics/dbt/pull/800) as well.
+Different data warehouses add levels to the hierarchy, and levers to your toolkit: 
+Snowflake allows an arbitrary number of logical databases; models in one BigQuery
+project can select data from another one, if you have access to both.
+
+A significant theme is the relationship between folder structure in a dbt repo
+and the schema structure of its target warehouse. We prefer setting high-level
+configurations in `dbt_project.yml`, where possible, and selecting only a
+specific set of models—`stg_` views and `fct_` tables—to expose in the
+tangible warehouse.
+
+This can be as simple as making all intermediate scripts
+a mix of ephemeral models (compiled SQL code that dbt injects in place of `ref()`
+statements) and tables in an unexposed schema. It can be as fancy as setting
+up Raw and Analytics logical databases in Snowflake, with a `salesforce.account`
+in each.
+
+### Staging Raw Data
+
+The staging model is the atomic unit of data modeling. Each model bears a 
+one-to-one relationship with the raw data table it represents: 
+`salesforce.account` becomes `stg_salesforce_account`. It has the same
+granularity, but the columns have been renamed, recast, or usefully
+reconsidered.
+
+Every raw data table should still flow through a base model of the form:
+```sql
+with source as (
+    
+    select * from raw_schema.raw_table
+    
+),
+
+renamed as (
+    
+    select
+    
+        firstname as first_name,
+        createdat::date as created_at,
+        ...
+    
+)
+
+select * from renamed
+```
+Depending on the quality of raw data, this base model could also be a staging
+model. A staged view might just as well warrant several models' worth of cleaning, 
+correcting, and categorizing. Raw data comes as it is; staging data is clean, 
+presentable, and ready for the curtain to rise.
+
+Staging models **can** have joins in them to field additional columns for context 
+or enrichment; add rows through unions and remove them through filters;
+deduplicate a natural key or hash together a 
+[surrogate one](https://github.com/fishtown-analytics/dbt-utils#surrogate_key-source); 
+but they should **not** have `group by` aggregations that would change their 

```

Comment:
*should not perform any operations 

it would also be a bad idea to, for example, join on a dates table. this would change the granularity.